Add a (local mode) Scalding Interpreter to Zeppelin

sriramkrishnan opened the pull request:
### What is this PR for?
Scalding (https://github.com/twitter/scalding) is a Scala library for writing MapReduce jobs.
This issue tracks the addition of a Scalding interpreter for Zeppelin. To keep this work incremental, this PR will focus on just a local mode implementation. The Hadoop mode can be a subsequent addition.

### What type of PR is it?
Feature

### Todos
* Addition of Hadoop mode for Scalding

### Is there a relevant Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-526

### How should this be tested?
Run the tests in: scalding/src/test/java/org/apache/zeppelin/scalding/ScaldingInterpreterTest.java

### Screenshots
<img width="1167" alt="scalding-example" src="https://cloud.githubusercontent.com/assets/1509691/11944979/8788d5c2-a7ff-11e5-9863-2a1216c51896.png">
<img width="1151" alt="scalding-screenshot" src="https://cloud.githubusercontent.com/assets/1509691/11944978/8787e3ec-a7ff-11e5-8383-456adb16b977.png">


### Questions:
* This could use documentation, which could just be the example in the screenshot. Where can I contribute that?

bzz commented on scalding/src/main/java/org/apache/zeppelin/scalding/ScaldingInterpreter.java:
```diff
@@ -0,0 +1,288 @@
+/**
+ * Scalding interpreter for Zeppelin. Based off the Spark interpreter code.
+ *
+ * @author sriramkrishnan
```
Would you be willing to remove that information from the comments, please?

It is not strictly documented yet, but Zeppelin so far, as many other ASF projects (Hadoop, Zookeeper, Avro, etc), do not encourage use of @author tags.

We definitely want and keep contributors credits, but we use git, JIRA and mailing list history, so nothing will be lost.

sriramkrishnan replied:
Sure, not a problem.

bzz commented on scalding/src/main/java/org/apache/zeppelin/scalding/ScaldingInterpreter.java:
```diff
@@ -0,0 +1,288 @@
+public class ScaldingInterpreter extends Interpreter {
+  Logger logger = LoggerFactory.getLogger(ScaldingInterpreter.class);
+
+  public static final List<String> NO_COMPLETION = new ArrayList<>();
```
A nit and does not block megeing, but this looks like [not a constant](https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names).

May be using some immutable collection implementation would be a good fit here (i.e Guava )

sriramkrishnan replied:
I would prefer not to add an additional dependency here. Also, not that it's an excuse, but this is basically copy/pasted from https://github.com/apache/incubator-zeppelin/blob/master/cassandra/src/main/java/org/apache/zeppelin/cassandra/CassandraInterpreter.java#L142 :).

bzz replied:
Sounds good. 
BTW you could wrap it in `Collections.unmodifiable` to archive the same result AND set the good example for others ;)

sriramkrishnan replied:
Alright, you got it!

bzz commented:
@sriramkrishnan could you please also add some docs for the new interpreter here http://zeppelin.incubator.apache.org/docs/0.6.0-incubating-SNAPSHOT/interpreter/ ?

Here are some details on how to [add it](http://zeppelin.incubator.apache.org/docs/0.6.0-incubating-SNAPSHOT/development/howtocontributewebsite.html).

prasadwagle commented:
Should we add ScaldingInterpreter to ZEPPELIN_INTERPRETERS in ZeppelinConfiguration.java?

tw-172-25-131-152 incubator-zeppelin (presto) $ git diff master -- zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java
diff --git a/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java b/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java
index 909345a..0c6e9f0 100755
--- a/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java
+++ b/zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java
@@ -415,6 +415,7 @@ public class ZeppelinConfiguration extends XMLConfiguration {
         + "org.apache.zeppelin.cassandra.CassandraInterpreter,"
         + "org.apache.zeppelin.geode.GeodeOqlInterpreter,"
         + "org.apache.zeppelin.postgresql.PostgreSqlInterpreter,"
+        + "org.apache.zeppelin.scalding.ScaldingInterpreter,"
         + "org.apache.zeppelin.kylin.KylinInterpreter"),
     ZEPPELIN_INTERPRETER_DIR("zeppelin.interpreter.dir", "interpreter"),
     ZEPPELIN_INTERPRETER_CONNECT_TIMEOUT("zeppelin.interpreter.connect.timeout", 30000),

sriramkrishnan commented:
@bzz thanks for the review. I will address your other comments and update the PR in a day or so.

@prasadwagle good point. I will add that too.

sriramkrishnan commented:
Addressed all PR comments, except for docs, license as @bzz has suggested, which I will do next. Build is also green now.

ps: I do need the hadoop-client jar as the Scalding REPL uses it even for local mode (and provided doesn't work there AFAICT).

sriramkrishnan commented:
Added docs.

@bzz I believe I have addressed all comments, except adding LICENSE info to zeppelin-distribution/src/bin_license/LICENSE. Do you happen to have a script for doing that already? I am dreading doing this by hand.

jongyoul commented:
I've tested it in my local machine and worked well. Thanks for this contribution.

sriramkrishnan commented:
Thanks @jongyoul!
 
@bzz I have gone ahead and manually added all the licenses I could find (phew!). There were some duplicate jars with different versions, which I didn't re-add.

I think I am mostly done with this PR. Let me know if this is good to merge, or if there is anything else I need to do.

Leemoonsoo commented:
Thanks @sriramkrishnan for great work.

I think following files
```
LICENSE-paranamer-2.3
LICENSE-protobuf-2.5.0
LICENSE-xmlenc-0.52
```
need to be moved from `/licenses` to `zeppelin-distribution/src/bin_license/licenses` directory.
And if `LICENSE-paranamer-2.3` and `LICENSE-protobuf-2.5.0` can have text format instead of html, that would be easier to read.




For `org.tukaani:xz:jar:1.0`, how about make a section, such as
```
========================================================================
For the org.tukaani:xz:jar:1.0
========================================================================
See licenses/LICENSE-tukaani-xy-1.0
```
and copy contents of `http://git.tukaani.org/?p=xz-java.git;a=blob;f=COPYING` into `zeppelin-distribution/src/bin_license/licenses/LICENSE-tukaani-xy-1.0`?




For `jgrapht-jdk1.6`, it became LGPL and EPL dual license since 0.9.x. But I'm not sure we can assume the license of 0.8.1 as EPL, since it was released with LGPL at that time. Can someone clarify?

Leemoonsoo commented on scalding/pom.xml:
```diff
@@ -0,0 +1,202 @@
+    <repository>
+      <id>conjars</id>
+      <name>Concurrent Maven Repo</name>
+      <url>http://conjars.org/repo</url>
```
It is possible to avoid, or at least disable this maven repo by default?
We need to avoid publish official binary package built with 3rd party maven repository.

sriramkrishnan replied:
All the Cascading jars are actually available from http://conjars.org/. I am not sure how I can disable or avoid this.

sriramkrishnan commented:
Thanks @Leemoonsoo I believe I have addressed all your comments, except for the jgrapht.

Note that the jgrapht here is thirdparty:jgrapht, which seems to be a version that Cascading has forked. I am inclined to leave it as it, but open to suggestions.

sriramkrishnan commented:
\*bump\*

What else do we need to do to get this merged? Thanks!

Leemoonsoo commented:
I have queried and waiting for answer about license of thirdparty:jgraph.

One of the fastest way to get this merged, making it optional like we did it for geode https://github.com/apache/incubator-zeppelin/pull/379/files.

That would avoid license problem and 3rd party maven repo. Then i think it can be merged.

Eventually, i hope https://issues.apache.org/jira/browse/ZEPPELIN-546 solve the problem.
i.e. Does not include scaling dependency in binary package of Zeppelin. but release scalding interpreter as a maven artifact. And then Zeppelin can load it, on runtime.

sriramkrishnan commented:
I am OK making the Scalding interpreter optional with a -Pscalding flag. Does that also mean I have to revert all my changes to the LICENSE file?

Leemoonsoo commented:
Yes it is, changes to LICENSE file and files that has been added to licenses directory need to be reverted. And you probably want to describe -Pscalding flag in https://github.com/apache/incubator-zeppelin/blob/master/README.md

Leemoonsoo commented:
fyi, i have got replied about license

>> moon soo Lee
>> Subject: Questions about license of jgrapht-jdk1.6:0.8.1 binary, like
>
> DEC 30, 2015  |  07:13AM PST 
> Ryan Desmond replied:
> Hi Moon,
>
> Chris has informed me that you can choose either license. That said, he also mentioned that you will need to speak with a lawyer for proper consultation on this issue.
>
> Best Regards,
> Ryan

sriramkrishnan commented:
Thanks @Leemoonsoo. So how should we proceed? Should I add it as LGPL?

Also, are you OK with using the conjars repo now? Or do you recommend that I add a -Pscalding flag. Would be great to get this merged either ways!

Leemoonsoo commented:
Apart from license for jgrapht, using 3rd party repo (conjars) need to be avoided. How about revert commits for the LICENSE / files added under licenses dir and add -Pscalding flag, for this pullrequest. After it is merged, create another issue for removing -Pscalding and continue to work. Would this way works for you?

sriramkrishnan commented:
@Leemoonsoo as you suggested, I have reverted commits for the LICENSE / files added under licenses dir, added -Pscalding flag, and updated the docs.

Are you OK with shipping/merging this PR now?

sriramkrishnan commented:
And we also have a green build now.

Leemoonsoo commented:
LGTM. @sriramkrishnan Thanks for the contribution!

sriramkrishnan commented:
Thanks @Leemoonsoo for the review. Could one of the committers please merge if there are no other objections?

Leemoonsoo commented:
I'm merging it, if there're no more discussions.

sriramkrishnan commented:
Thanks!

sriramkrishnan commented:
Filed https://issues.apache.org/jira/browse/ZEPPELIN-555 to track addition of Hadoop mode to the Scalding interpreter. I may pick it up in my "free time", but I have added a description there if anyone wants to tackle it.
